runtime/memprof.c

```diff
@@ -220,17 +232,30 @@ static value do_callback(tag_t tag, intnat wosize, int32_t occurences,
 }
 
 void caml_memprof_set_suspended(int new_suspended) {
-  caml_memprof_suspended = new_suspended;
+  if(caml_memprof_suspended != new_suspended) {
+    caml_memprof_suspended = new_suspended;
+    caml_memprof_renew_minor_sample();
+  }
```

Comment:
There are two ways you can imagine for suspending/resuming sampling:
1- (virtually) Setting lambda to 0
2- Ignoring all the samples that we generate

What I implemented is 1-: This means that, when suspended, everything happens as if `lambda=0`, i.e., `caml_memprof_young_trigger` becomes equal to `caml_young_alloc_start`, and `caml_memprof_track_young` is never entered. But then suspending/resuming involve  changing the value of lambda, which requires renewing the next minor sample. Note, however,  that renewing in the suspended state is very cheap (no random sampling is performed), so the actual renewing is only performed once.

The other possibility would be to implement 2-. I don't think the performance would be significantly different, but probably the code is going to be simpler. Let me try.

Reply:
I see. So if you had two functions as I suggested above, but with the current approach of "morally setting lambda to 0", they could look like

```
void caml_memprof_pause_sampling(void) {
  if (caml_memprof_suspended) return;
  caml_memprof_suspended = 1;
  set_trigger(caml_young_alloc_start);
}

void caml_memprof_resume_sampling(void) {
  if (!caml_memprof_suspended) return;
  caml_memprof_suspended = 0;
  caml_memprof_renew_minor_sample();
}
```

I can see now why the two are equivalent and your approach is more factorized, but at first it was a bit confusing to see logic to renew the sample when *suspending* sampling (actually it does not renew but, on the contrary, it removes the trigger).

Another implementation option one could use it to save the current `lambda` value in a `caml_memprof_suspended_lambda`, and actually set `lambda` to `0`, instead of keeping a separate boolean variable. This makes the check-sites slightly simpler.

If I understand correctly, the idea of the second implementation you are talking about is to check for `caml_memprof_suspended` in the `caml_memprof_track_*` functions and return directly in that case. (That does look very simple.)

Reply:
> If I understand correctly, the idea of the second implementation you are talking about is to check for caml_memprof_suspended in the caml_memprof_track_* functions and return directly in that case. (That does look very simple.)

Indeed. I pushed that new implementation.